npm/private/npm_import.bzl

```diff
@@ -39,208 +42,286 @@ _LINK_JS_PACKAGE_LOADS_TMPL = """\
+VERSION = "{version}"
+_ROOT_PACKAGE = "{root_package}"
 
-# buildifier: disable=bzl-visibility
-load("@aspect_rules_js//npm/private:npm_package_internal.bzl", _npm_package_internal = "npm_package_internal")\
-"""
-
-_LINK_JS_PACKAGE_TMPL = """
 # Generated npm_package_store targets for npm package {package}@{version}
 # buildifier: disable=function-docstring
 def npm_imported_package_store(name):
+    _npm_imported_package_store(
+        name = name,
+        package = PACKAGE,
+        version = VERSION,
+        root_package = _ROOT_PACKAGE,
+        deps = {deps},
+        ref_deps = {ref_deps},
+        lc_deps = {lc_deps},
+        dev = {dev},
+        has_lifecycle_build_target = {has_lifecycle_build_target},
+        transitive_closure_pattern = {transitive_closure_pattern},
+        npm_package_target = "{npm_package_target}",
+        package_store_name = "{package_store_name}",
+        lifecycle_hooks_env = {lifecycle_hooks_env},
+        lifecycle_hooks_execution_requirements = {lifecycle_hooks_execution_requirements},
+        use_default_shell_env = {use_default_shell_env},
+        exclude_package_contents = {exclude_package_contents},
+    )
+"""
+
+def npm_imported_package_store(
+        name,
+        package,
+        version,
+        root_package,
+        deps,
+        ref_deps,
+        lc_deps,
+        dev,
+        has_lifecycle_build_target,
+        transitive_closure_pattern,
+        npm_package_target,
+        package_store_name,
+        lifecycle_hooks_env,
+        lifecycle_hooks_execution_requirements,
+        use_default_shell_env,
+        exclude_package_contents):
     bazel_package = native.package_name()
-    root_package = "{root_package}"
     is_root = bazel_package == root_package
     if not is_root:
-        msg = "No store links in bazel package '%s' for npm package npm package {package}@{version}. This is neither the root package nor a link package of this package." % bazel_package
+        msg = "No store links in bazel package '{bazel_package}' for npm package npm package {package}@{version}. This is neither the root package nor a link package of this package.".format(
+            bazel_package = bazel_package,
+            package = package,
+            version = version,
+        )
         fail(msg)
-    if not name.endswith("/{package}"):
-        msg = "name must end with one of '/{package}' when linking the store in package '{package}'; recommended value is 'node_modules/{package}'"
+
+    suffix = "/" + package
+    if not name.endswith(suffix):
+        msg = "name must end with '/{package}' when linking the store in package '{package}'; recommended value is 'node_modules/{package}'".format(
+            package = package,
+        )
         fail(msg)
-    link_root_name = name[:-len("/{package}")]
+    link_root_name = name[:-len(suffix)]
 
-    deps = {deps}
-    ref_deps = {ref_deps}
+    deps = {k.format(link_root_name = link_root_name): v for k, v in deps.items()}
```

Comment:
It would probably be good to avoid `.items()` create giant new arrays?

Reply:
i'd love your opinion here. the problem is that `link_root_name` is not known at the time these dicts are constructed, only later. is it possible to compute `link_root_name` in the module extension and fully materialize these?

alternately, we could defer this expansion until the point where the `deps` are actually used, but i wasn't sure if there was a nice single point for that or not. 

I figured you know this better, wdyt?

Reply:
my preference would be to fully materialize at extension time if that's possible

Reply:
Ideally it could all be avoided until later.

Is this where dropping the `link_root_name` param would help?

Reply:
Ok I will try to trace it and see if we can defer, but yeah if we can drop lonk_root_name this becomes a non-issue

Reply:
@jbedard I think it's not possible - the `deps` are passed to `_npm_package_store_impl` at which point they need to be a `label_keyed_string_dict` so we must construct the label names before instantiating the rule. But dropping `link_root_name` would certainly make this smoother